servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -450,25 +374,47 @@ public void completeFile(long fileId)
         }
       }
 
-      completeFileInternal(fileInode.getBlockIds(), fileId, fileLength, false, opTimeMs);
+      // If the file is both persisted and stored in memory, the memory footprint and the UFS file
+      // length should match.
+      if (fileInode.isPersisted() && inMemoryLength != 0 && options.getLength() != inMemoryLength) {
+        throw new SuspectedFileSizeException("Inconsistent file length: Tachyon " + inMemoryLength
+            + " UFS " + options.getLength());
+      }
+
+      // If the file is persisted, its length is determined by UFS. Otherwise, its length is
+      // determined by its memory footprint.
+      long length = fileInode.isPersisted() ? options.getLength() : inMemoryLength;
```

Comment:
It seems like the logic is really tricky, on the client side we only set `options.length` if its persisted to ufs, and here we switch on `isPersisted` to determine if we should use `options.length`. Perhaps changing `length` to `ufsLength` or something would make it more apparent what the option is used for?